### ISSUE TITLE
Strange results from using pipes

### ISSUE THREAD

**Issue by ewelot** [libvips/libvips#1916]
I am experimenting with I/O redirections using the following simple program `test.py`:
```
#!/usr/bin/python3

import sys
import pyvips

if(len(sys.argv)>1 and sys.argv[1]!='-'):
    image = pyvips.Image.new_from_file(sys.argv[1])
else:
    source = pyvips.Source.new_from_descriptor(sys.stdin.fileno())
    image = pyvips.Image.new_from_source(source, "")

print(image.getpoint(128, 169))
image.ppmsave("x.ppm")
```
with a certain 16bit PPM for input (see attachment).
Running the program in two different ways produces different results:
```
$ test.py < test.ppm
[51505.0, 51505.0, 51505.0]
$ cat test.ppm | test.py
[12745.0, 12745.0, 12745.0]
```
Comparing input and output images you will recognize that some output pixels get strangely modified when running the piped command.
Any idea what is wrong with this?

(Example image test.ppm in [test.zip](https://github.com/libvips/libvips/files/5635895/test.zip))

**Comment by jcupitt** [comment]
Yes, it looks like there's some kind of bizarre bug here. I'll investigate. Thanks for the report!

**Comment by jcupitt** [comment]
OK, I think this is fixed in 8.10 and master -- there were a couple of small errors in 16-bit ppm handling. 

Thanks again for reporting this!
